### ISSUE TITLE
Overhead from ptrace detaching/reattaching when using the work stealing scheduler

### ISSUE THREAD

**Issue by sporksmith** [shadow/shadow#1182]
Linux only allows the thread (task) that is directly ptrace-attached to a target task to issue ptrace commands to that task. This becomes a problem for the work-stealing scheduler, which migrates hosts amongst worker threads on the fly. Right not we work around this by stopping and ptrace-detaching from all plugin tasks at the start of a scheduler round, and then only reattaching just as the worker starts executing the plugin.

Detaching requires a single PTRACE_DETACH operation, with SIGSTOP set to prevent the task from executing while detached. Reattaching requires a PTRACE_ATTACH operation, a waitpid for the next state (which will be a signal-stop from the SIGSTOP), and then a PTRACE_SETOPTIONS operation.

We haven't yet measured how much overhead this contributes to "real" workloads in the work-stealing-scheduler, but from measuring the overhead of extra reattaching in the o(n)-waitpid-workarounds, we can infer that it's as much as [40%](https://github.com/shadow/shadow/issues/1134#issuecomment-798992292) in the ~worst case where we only process a single syscall per host in between having to detach and reattach again.

Some options for reducing or eliminating this overhead:

* **Multithreaded ptrace kernel patch**. Patch the kernel to allow any thread in the tracing process to issue ptrace operations. This is probably the ideal solution both performance-wise and overall-complexity-wise. Unfortunately it is likely to require adding some extra complexity and an extra mutex to the task structure in the Linux kernel, which may be a hard sell to upstream without more mainstream use-cases. [LKML thread](https://lkml.org/lkml/2021/3/10/1398)
* **Tracing threads**. Use one or more dedicated "tracing threads" whose only purpose is to issue ptrace operations. This is fairly simple to implement, but not a clear performance win since each ptrace-operation will require a "thread hop" back and forth. If all CPUs are used for workers+plugins, the worker would need to be scheduled off of the current core, the tracer thread scheduled on to issue the operation, and then the worker thread scheduled back on. We could potentially avoid some of that overhead by leaving some CPUs free for the tracer threads to run on concurrently, but it probably wouldn't be worth "losing" those CPUs. Requires o1-waitpid-patch to avoid quadratic scaling.
* **Host threads**. Associate a "host thread" with each host. When a worker "executes" a host, it'd invoke that thread (e.g. via a semaphore) instead of calling into the host code directly. Since the host thread is what would be ptrace-attached, and hosts are the granularity at which we schedule work across workers/CPUs, we'd never have to reattach. This has much less thread-hopping than using a small number of tracer threads, but requires many more total threads, which add to memory overhead in kernel data structures and for each thread's stack (though we could probably clamp this down much smaller than the default 8 MB stacks). This approach would also largely avoid quadratic scaling from waitpid *without* requiring the o-1-waitpid kernel patch, since each host typically only has a small number of plugin tasks.
* **Lazy detach**. Currently we detach all plugins in between scheduler rounds so that they can move if they need to, which means we pay the detach+reattach overhead even if it *doesn't* move. We could mitigate this by adding a mechanism for one worker to send a request to another worker to detach from a plugin so that it can be migrated. However this means we'd need to add some mechanism for the worker threads to "listen" for such requests - probably by polling a queue in between plugin-invocations. In theory this could result in arbitrarily high latency for such requests to be serviced (when executing a plugin that does a lot of processing between syscalls), but could be ok in practice.

**Comment by sporksmith** [comment]
Some ideas for measuring the reattach overhead (from discussion with @robgjansen)

* Add an option to do the reattach in host mode as well, and measure the performance difference with/without it. This should work well for a "balanced" workload, but might have some confusing interactions with the effect of idle workers in unbalanced workloads, since the reattaching would be "serialized" over the smaller number of workers doing the work.
* Add explicit timers around the detach and reattach operation, and then measure how much time is spent there in a tor experiment with the work stealing scheduler.
* Look at a `perf` report (or some other tracing tool such as LTTng) for a tor simulation with the work stealing scheduler. This would be a bit less precise than explicit timers, but we could do it without modifying the code.

I'm inclined to try the `perf` route first since it doesn't require code modification, we can do it on a "real" workload, and it's probably precise enough.

**Comment by sporksmith** [comment]
Some implementation ideas for the **host threads** strategy (from discussion with @robgjansen):

We may be able to encapsulate a prototype of this strategy as a new scheduler-strategy. Instead of a new class of "host" threads, we'd use as many worker threads as hosts, and use a modified version of the "host_single" strategy. The scheduler-strategy would block on a semaphore waiting for an available CPU, and when it wakes adjust CPU affinity appropriately. When it finishes with a host it should relax its affinity again so that it can be re-woken on any CPU, post to the semaphore so that the next worker can wake up, and then wait on the semaphore to be woken up again the next round.

This approach would likely have more CPU migrations than the work-stealing strategy, since each time a host is awoken, it'd come back up on an arbitrary CPU.

Actually going this route we wouldn't even *necessarily* need to have fully as many worker threads as hosts, giving us some flexibility in trading off # of threads and granularity of work. Since there'd be no work stealing in this design though, getting "unlucky" and putting two of the most expensive hosts on the same worker could significantly hurt performance.

**Comment by robgjansen** [comment]
I'm wondering if there is something smart we can do here to limit the number of CPU migrations. If we were able to control which worker gets woken up when a given CPU is available, then we could prioritize the next worker that gets woken up to one whose affinity matches the available CPU. That way, we would only have CPU migrations when there are no other workers waiting that have the affinity of the available CPU.

**Comment by sporksmith** [comment]
I think we can do it with a synchronized queue for each CPU, containing worker ids or pointers, and another set for "next" round. Something like:

* When finishing work for the current round, try popping a worker from queue for this CPU. If the queue is empty, scan the other queues. If we successfully pop from another queue, adjust the CPU affinity for that worker.
* Wake the worker we popped, if there was one.
* Go to sleep on the "next round" queue for the CPU we just ran on

**Comment by sporksmith** [comment]
Some ad-hoc measurements of thread overhead for viability:

As roughly a lower bound, Rust threads that do nothing consume about 35 kB each on my machine. Measured with:

```
use std::io::{BufRead, BufReader};
use std::sync::{Arc,Barrier};
use std::fs::File;
use std::thread;

fn get_available_mem() -> usize {
    let file = File::open("/proc/meminfo").unwrap();
    let reader = BufReader::new(file);
    for line in reader.lines() {
        // println!("{}", line.unwrap());
        let line = line.unwrap();
        if line.starts_with("MemAvailable:")  {
            let fields = line.split_whitespace();
            let avail = fields.skip(1).next().unwrap();
            return avail.parse().unwrap();
        }
    }
    panic!("Didn't find MemAvailable");
}

fn main() {
    let n : usize = std::env::args().skip(1).next().unwrap().parse().unwrap();
    let available0 = get_available_mem();
    println!("{} kB available startup", available0);

    println!("\nSpawning {} threads", n);
    let started = Arc::new(Barrier::new(n+1));
    let done = Arc::new(Barrier::new(n+1));
    for _ in 0..n {
        let started = started.clone();
        let done = done.clone();
        thread::spawn(move || {
            // Wait for all threads to spawn, implicitly notifying main thread when all have
            // spawned.
            started.wait();
            // Wait for main thread to finish
            done.wait();
        });
    }
    // Wait for all threads to spawn
    started.wait();

    let available1 = get_available_mem();
    println!("{} kB available with threads", available1);
    println!("{} kB / thread", (available0-available1)/n);

    // We could call done.wait here to allow threads to finish, but might as well just exit and let
    // the OS kill them.
}
```

Shadow worker threads are likely to consume more, since they will actually use their stacks and have a bit of thread local storage.

I measured using the minimal tor simulation, and setting workers to 1, 1000, and 2000. In a separate terminal I used a small script to sample the available once per second. After the simulation was done I took the difference between the lowest and highest values to be Shadow's memory usage.

Using w=1 as the baseline (accounting for all the memory Shadow uses beyond for the threads themselves), I got 163 kB/worker @ 1000 workers and 46 kB/worker @ 2000 workers. It may be worth trying this again on a dedicated machine rather than my dev machine, but that at least gives us a ballpark estimate of 40-160 kB/worker.

Shell bit for sampling:
```
while true ; do grep Available /proc/meminfo | grep -o -E '[[:digit:]]+' >> mem; sleep 1; done
```

**Comment by sporksmith** [comment]
Closing, as this is now addressed in #1238, which effectively implements the "host threads" solution. We can now set the number of worker threads as high as we like - e.g. 1 per host, and independently configure how many to actually allow to execute at once via `--max-parallelism`. When configured this way, and using the Host scheduler, work is shared at Host granularity (as with the work stealing scheduler), but without needing any ptrace detach or reattach.

So far the overhead from the extra threads seems not too bad, though if we ever wanted to reduce it, we could still consider using fewer threads and adding support for "lazy detach", so that we could still steal Hosts.
